core/trino-main/src/main/java/io/trino/operator/output/UnnestingPositionsAppender.java

```diff
@@ -59,12 +69,15 @@ public void appendRle(Block block, int rlePositionCount)
         if (rlePositionCount == 0) {
             return;
         }
+        dictionaryBlockBuilder.flushDictionary(delegate);
         delegate.appendRle(block, rlePositionCount);
     }
 
     @Override
     public void append(int position, Block source)
     {
+        dictionaryBlockBuilder.flushDictionary(delegate);
```

Comment:
Why is dictionary flushed in every case here but in the other `append` it is flushed only for non-dictionary cases

Reply:
I think we flush in every case other than we have the same dictionary (`dictionaryBlockBuilder.canAppend` to be precise).
We could also do it here as well at a cost of `!closed && (dictionary == this.dictionary || this.dictionary == null)` for every row. I will try to benchmark how much this impacts performance of the row by row processing